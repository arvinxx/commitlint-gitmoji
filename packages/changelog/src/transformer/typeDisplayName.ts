import type { CommitTypes } from '@gitmoji/commit-types';
import { merge } from 'lodash';

export interface DisplayNameOptions {
  withEmoji?: boolean;
  language?: 'en-US' | 'zh-CN' | 'mix';
  customTypeMap?: { [key in CommitTypes]?: CustomTypeNameMap };
}

export interface CustomTypeNameMap {
  emoji?: string;
  'en-US'?: string;
  'zh-CN'?: string;
  subtitle?: string;
}

interface TypeNameMap {
  emoji: string;
  'en-US': string;
  'zh-CN': string;
  subtitle: string;
}

export const typeMap: Record<Exclude<CommitTypes, 'wip'>, TypeNameMap> = {
  feat: {
    emoji: '✨',
    'en-US': 'Features',
    'zh-CN': '新特性',
    subtitle: "What's improved",
  },
  fix: {
    emoji: '🐛',
    'en-US': 'Bug Fixes',
    'zh-CN': '修复',
    subtitle: "What's fixed",
  },
  build: {
    emoji: '👷',
    'en-US': 'Build System',
    'zh-CN': '构建系统',
    subtitle: 'Build system',
  },
  chore: {
    emoji: '🎫',
    'en-US': 'Chores',
    'zh-CN': '杂项',
    subtitle: 'Chores',
  },
  ci: {
    emoji: '🔧',
    'en-US': 'Continuous Integration',
    'zh-CN': '持续集成',
    subtitle: 'Continuous integration',
  },
  docs: {
    emoji: '📝',
    'zh-CN': '文档',
    'en-US': 'Documentation',
    subtitle: 'Documentation',
  },
  test: {
    emoji: '✅',
    'zh-CN': '测试',
    'en-US': 'Tests',
    subtitle: 'Tests',
  },
  perf: {
    emoji: '⚡',
    'en-US': 'Performance Improvements',
    'zh-CN': '性能优化',
    subtitle: 'Performance improvements',
  },
  refactor: {
    emoji: '♻',
    'en-US': 'Code Refactoring',
    'zh-CN': '重构',
    subtitle: 'Code refactoring',
  },
  revert: {
    emoji: '⏪',
    'zh-CN': '回滚',
    'en-US': 'Reverts',
    subtitle: 'Reverts',
  },
  style: {
    emoji: '💄',
    'en-US': 'Styles',
    'zh-CN': '样式',
    subtitle: 'Styles',
  },
};

export const defineTypeMap = (
  customTypeMap: { [key in CommitTypes]?: CustomTypeNameMap },
): Record<Exclude<CommitTypes, 'wip'>, TypeNameMap> => {
  if (!customTypeMap) return typeMap;
  return merge(typeMap, customTypeMap);
};

export const getDisplayName = (
  type: CommitTypes | string,
  options: DisplayNameOptions = {},
): string => {
  const { withEmoji = true, language = 'en-US' } = options;

  const diplayTypeMap = defineTypeMap(options.customTypeMap);

  if (type in diplayTypeMap) {
    const item = diplayTypeMap[type];
    const { emoji } = item;
    return `${withEmoji ? `${emoji} ` : ''}${
      language === 'mix' ? [item['en-US'], item['zh-CN']].join(' | ') : item[language]
    }`;
  }

  return type;
};
