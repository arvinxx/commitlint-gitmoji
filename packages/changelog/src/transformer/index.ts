import type { CommitTypes } from '@gitmoji/commit-types';
import types from '@gitmoji/commit-types';
import type { Context } from 'conventional-changelog-writer';
import type { Commit } from 'conventional-commits-parser';
import pangu from 'pangu';
import type { CustomConfig } from '../customConfig';
import { scopeMapDisplayName } from './scopeMapDisplayName';
import { getDisplayName } from './typeDisplayName';

const capitalizeFirstLetter = (str: string): string => {
  const firstLetter = String(str).slice(0, 1).toUpperCase();
  const remainingStr = String(str).slice(1);
  return firstLetter + remainingStr;
};

export default (customConfig: CustomConfig) => (commit: Commit, context: Context) => {
  let discard = true;
  const issues = [];

  commit.notes.forEach((note) => {
    note.title = `${customConfig?.withEmoji === false ? '' : '💥 '}BREAKING CHANGES`;

    discard = false;
  });

  let displayTypes = types;

  if (customConfig.displayTypes) {
    displayTypes = customConfig.displayTypes;
  }

  if (!displayTypes.includes(<CommitTypes>commit.type) && discard) return;

  // 修改 type 标题
  commit.type = getDisplayName(commit.type, {
    language: customConfig.titleLanguage,
    withEmoji: customConfig.withEmoji,
  });

  /** * 处理 scope ** */
  if (commit.scope === '*') {
    commit.scope = '';
  }

  if (customConfig.displayScopes) {
    if (!customConfig.displayScopes?.includes(commit.scope)) return;
  }

  if (customConfig.scopeDisplayName) {
    commit.scope = scopeMapDisplayName(commit.scope, customConfig.scopeDisplayName);
  }

  if (typeof commit.hash === 'string') {
    commit.hash = commit.hash.substring(0, 7);
  }

  if (typeof commit.subject === 'string') {
    let url = context.repository
      ? `${context.host}/${context.owner}/${context.repository}`
      : context.repoUrl;
    if (url) {
      url = `${url}/issues/`;
      // Issue URLs.
      commit.subject = commit.subject.replace(/#([0-9]+)/g, (_, issue) => {
        issues.push(issue);
        return `[#${issue}](${url}${issue})`;
      });
    }
    if (context.host) {
      // User URLs.
      commit.subject = commit.subject.replace(
        /\B@([a-z0-9](?:-?[a-z0-9/]){0,38})/g,
        (_, username) => {
          if (username.includes('/')) {
            return `@${username}`;
          }

          return `[@${username}](${context.host}/${username})`;
        },
      );
    }
  }

  // remove references that already appear in the subject
  commit.references = commit.references.filter((reference) => {
    return issues.indexOf(reference.issue) === -1;
  });

  // format
  if (commit.authorNameEncode) commit.authorNameEncode = encodeURIComponent(commit.authorName);
  if (commit.subject) commit.subject = pangu.spacing(capitalizeFirstLetter(commit.subject));

  return commit;
};
