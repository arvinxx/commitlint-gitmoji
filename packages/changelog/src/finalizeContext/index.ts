import type { Context } from 'conventional-changelog-writer';
import { CustomConfig } from '../customConfig';
import { typeMap } from '../transformer/typeDisplayName';
export default (customConfig: CustomConfig) => (context: Context): Context => {
  const subCommitScope = customConfig?.scopeDisplayName?.['*'] || '';
  const authors = {};
  context.commitGroups = context.commitGroups.map((item) => {
    const subtitle = Object.values(typeMap).find(
      (i) =>
        item.title.includes(i['emoji']) ||
        item.title.includes(i['en-US']) ||
        item.title.includes(i['zh-CN']),
    ).subtitle;
    let group;
    let commits = item.commits.sort((a, b) => {
      if (a.scope === subCommitScope && b.scope === subCommitScope) {
        return 0;
      } else if (a.scope === subCommitScope) {
        return 1;
      } else if (b.scope === subCommitScope) {
        return -1;
      } else {
        return 0;
      }
    });
    commits = commits.map((c, index) => {
      if (group !== c.scope) {
        group = c.scope;
        c.first = true;
      } else {
        c.first = false;
      }
      if (!commits[index + 1] || group !== commits[index + 1].scope) {
        c.last = true;
      } else {
        c.last = false;
      }
      if (c.authorNameEncode && !authors[c.authorNameEncode]) {
        authors[c.authorNameEncode] = {
          authorName: c.authorName,
          authorEmail: c.authorEmail,
          authorNameEncode: c.authorNameEncode,
        };
      }
      return c;
    });
    return {
      title: item.title,
      subtitle,
      commits,
    };
  });
  context.authors = Object.values(authors);
  return context;
};
