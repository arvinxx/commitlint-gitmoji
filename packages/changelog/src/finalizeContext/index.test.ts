import { typeMap } from '../transformer/typeDisplayName';
import finalizeContext from './index';

describe('finalizeContext', () => {
  const customConfig = {
    scopeDisplayName: {
      '*': 'all',
    },
    customTypeMap: {},
  };
  const context = {
    commitGroups: [
      {
        title: '✨ Features',
        commits: [
          {
            hash: '1234',
            subject: 'test commit',
            scope: 'test',
            title: '✨ Features',
            authorName: 'canisminor1990',
            authorEmail: 'i@canisminor.cc',
            authorAvatar: 'https://avatars.githubusercontent.com/u/17870709?v=4',
            authorUrl: 'https://api.github.com/users/canisminor1990',
          },
        ],
      },
    ],
    authors: [],
  };

  it('should transform commitGroups correctly', () => {
    const transformedContext = finalizeContext(customConfig)(context);
    expect(transformedContext).toMatchSnapshot();
  });

  it('should set subCommitScope correctly when it is defined in customConfig', () => {
    const customConfigWithSubCommitScope = {
      ...customConfig,
      scopeDisplayName: {
        '*': 'all',
        test: 'test',
      },
    };
    const transformedContext = finalizeContext(customConfigWithSubCommitScope)(context);
    expect(transformedContext.commitGroups[0].commits[0].first).toBe(true);
  });

  it('should set subCommitScope to null when it is not defined in customConfig', () => {
    const transformedContext = finalizeContext(customConfig)(context);
    expect(transformedContext.commitGroups[0].commits[0].first).toBe(true);
  });

  it('should set subtitle correctly when title contains emoji', () => {
    const contextWithTitleEmoji = {
      ...context,
      commitGroups: [
        {
          title: '✨ Features',
          commits: [],
        },
      ],
    };
    const transformedContext = finalizeContext(customConfig)(contextWithTitleEmoji);
    expect(transformedContext.commitGroups[0].subtitle).toBe("What's improved");
  });

  it('should set subtitle correctly when title contains en-US', () => {
    const contextWithTitleEnUS = {
      ...context,
      commitGroups: [
        {
          title: 'Features',
          commits: [],
        },
      ],
    };
    const transformedContext = finalizeContext(customConfig)({
      ...contextWithTitleEnUS,
      commitGroups: [
        {
          ...contextWithTitleEnUS.commitGroups[0],
          title: `Features ${typeMap['feat']['en-US']}`,
        },
      ],
    });
    expect(transformedContext.commitGroups[0].subtitle).toBe("What's improved");
  });

  it('should set subtitle correctly when title contains zh-CN', () => {
    const contextWithTitleZhCN = {
      ...context,
      commitGroups: [
        {
          title: 'Features',
          commits: [],
        },
      ],
    };
    const transformedContext = finalizeContext(customConfig)({
      ...contextWithTitleZhCN,
      commitGroups: [
        {
          ...contextWithTitleZhCN.commitGroups[0],
          title: `Features ${typeMap['feat']['zh-CN']}`,
        },
      ],
    });
    expect(transformedContext.commitGroups[0].subtitle).toBe("What's improved");
  });

  it('should sort commits correctly when subCommitScope is defined in customConfig', () => {
    const customConfigWithSubCommitScope = {
      ...customConfig,
      scopeDisplayName: {
        '*': 'all',
        test: 'test',
      },
    };
    const contextWithMultipleCommits = {
      ...context,
      commitGroups: [
        {
          title: '✨ Features',
          commits: [
            {
              hash: '1234',
              subject: 'test commit 1',
              scope: 'test',
              title: '✨ Features',
              authorName: 'canisminor1990',
              authorEmail: 'i@canisminor.cc',
              authorAvatar: 'https://avatars.githubusercontent.com/u/17870709?v=4',
              authorUrl: 'https://api.github.com/users/canisminor1990',
            },
            {
              hash: '5678',
              subject: 'test commit 2',
              scope: 'test',
              title: '✨ Features',
              authorName: 'canisminor1990',
              authorEmail: 'i@canisminor.cc',
              authorAvatar: 'https://avatars.githubusercontent.com/u/17870709?v=4',
              authorUrl: 'https://api.github.com/users/canisminor1990',
            },
          ],
        },
      ],
    };
    const transformedContext = finalizeContext(customConfigWithSubCommitScope)(
      contextWithMultipleCommits,
    );
    expect(transformedContext.commitGroups[0].commits[0].first).toBe(true);
    expect(transformedContext.commitGroups[0].commits[1].first).toBe(false);
  });

  it('should set first and last correctly when commits have the same scope', () => {
    const contextWithMultipleCommits = {
      ...context,
      commitGroups: [
        {
          title: '✨ Features',
          commits: [
            {
              hash: '1234',
              subject: 'test commit 1',
              scope: 'test',
              title: '✨ Features',
              authorName: 'canisminor1990',
              authorEmail: 'i@canisminor.cc',
              authorAvatar: 'https://avatars.githubusercontent.com/u/17870709?v=4',
              authorUrl: 'https://api.github.com/users/canisminor1990',
            },
            {
              hash: '5678',
              subject: 'test commit 2',
              scope: 'test',
              title: '✨ Features',
              authorName: 'canisminor1990',
              authorEmail: 'i@canisminor.cc',
              authorAvatar: 'https://avatars.githubusercontent.com/u/17870709?v=4',
              authorUrl: 'https://api.github.com/users/canisminor1990',
            },
          ],
        },
      ],
    };
    const transformedContext = finalizeContext(customConfig)(contextWithMultipleCommits);
    expect(transformedContext.commitGroups[0].commits[0].first).toBe(true);
    expect(transformedContext.commitGroups[0].commits[0].last).toBe(false);
    expect(transformedContext.commitGroups[0].commits[1].first).toBe(false);
    expect(transformedContext.commitGroups[0].commits[1].last).toBe(true);
  });

  it('should set author correctly when authorAvatar is not empty', () => {
    const contextWithAuthorNameEncode = {
      ...context,
      commitGroups: [
        {
          title: '✨ Features',
          commits: [
            {
              hash: '1234',
              subject: 'test commit',
              scope: 'test',
              title: '✨ Features',
              authorName: 'canisminor1990',
              authorEmail: 'i@canisminor.cc',
              authorAvatar: 'https://avatars.githubusercontent.com/u/17870709?v=4',
              authorUrl: 'https://api.github.com/users/canisminor1990',
            },
          ],
        },
      ],
    };
    const transformedContext = finalizeContext(customConfig)(contextWithAuthorNameEncode);
    expect(transformedContext.authors).toEqual([
      {
        authorAvatar: 'https://avatars.githubusercontent.com/u/17870709?v=4',
        authorEmail: 'i@canisminor.cc',
        authorName: 'canisminor1990',
        authorUrl: 'https://api.github.com/users/canisminor1990',
      },
    ]);
  });

  it('should not set author when authorAvatar is empty', () => {
    const contextWithoutAuthorNameEncode = {
      ...context,
      commitGroups: [
        {
          title: '✨ Features',
          commits: [
            {
              hash: '1234',
              subject: 'test commit',
              scope: 'test',
              title: '✨ Features',
              authorName: 'canisminor1990',
              authorEmail: 'i@canisminor.cc',
            },
          ],
        },
      ],
    };
    const transformedContext = finalizeContext(customConfig)(contextWithoutAuthorNameEncode);
    expect(transformedContext.authors).toEqual([]);
  });
});
