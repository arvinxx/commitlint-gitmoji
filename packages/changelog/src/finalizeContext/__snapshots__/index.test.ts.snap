// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`finalizeContext should transform commitGroups correctly 1`] = `
{
  "authors": [
    {
      "authorEmail": "test@test.com",
      "authorName": "test",
      "authorNameEncode": "test",
    },
  ],
  "commitGroups": [
    {
      "commits": [
        {
          "authorEmail": "test@test.com",
          "authorName": "test",
          "authorNameEncode": "test",
          "first": true,
          "hash": "1234",
          "last": true,
          "scope": "test",
          "subject": "test commit",
          "title": "✨ Features",
        },
      ],
      "subtitle": "What's improved",
      "title": "✨ Features",
    },
  ],
}
`;

exports[`test transformCommitGroups should transform commitGroups correctly 1`] = `
{
  "authors": [
    {
      "authorEmail": "test@test.com",
      "authorName": "test",
      "authorNameEncode": "test",
    },
  ],
  "commitGroups": [
    {
      "commits": [
        {
          "authorEmail": "test@test.com",
          "authorName": "test",
          "authorNameEncode": "test",
          "first": true,
          "hash": "1234",
          "last": true,
          "scope": "test",
          "subject": "test commit",
          "title": "✨ Features",
        },
      ],
      "subtitle": "What's improved",
      "title": "✨ Features",
    },
  ],
}
`;
